src/query/storage/cache/redis_cache.go

```diff
@@ -35,9 +94,186 @@ func NewRedisCache(redisAddress string, logger *zap.Logger) *RedisCache {
 		return nil
 	}
 	logger.Info("Connection to Redis established", zap.String("address", redisAddress))
-	return &RedisCache{
+	cache := &RedisCache{
 		client:       pool,
 		redisAddress: redisAddress,
 		logger:       logger,
+		cacheMetrics: NewCacheMetrics(scope),
+	}
+	return cache
+}
+
+// Given a fetch query, converts it into a key for Redis
+func keyEncode(query *storage.FetchQuery) string {
+	res := make([]string, len(query.TagMatchers))
+	for i, m := range query.TagMatchers {
+		res[i] = m.String()
+	}
+	// Sort to guarantee we have the same order every time
+	sort.Strings(res)
+
+	label_key := strings.Join(res, ";")
+	queryRange := int64(query.End.Sub(query.Start).Seconds())
+	// Key becomes {labels}::{endTime}::{duration}
+	// For example, a key might look like
+	// "fieldA=\"1\";fieldB=\"2\"::1659459470::300"
+	return fmt.Sprintf("%s::%d::%d", label_key, query.End.Unix(), queryRange)
+}
+
+// Given a result, encodes it into string (or "" if error)
+// Encodes an empty PromResult to {} to differentiate from ""
+//
+// We only encode the QueryResult portion of the PromResult struct
+// We omit the metadata since we do not use it in computation
+// and also this metadata also isn't accurate since we aren't getting it from M3DB
+func resultEncode(result *storage.PromResult) (string, error) {
+	if len(result.PromResult.Timeseries) == 0 {
+		return EmptyResult, nil
+	}
+	// Custom Marshal function for QueryResults
+	res, err := result.PromResult.Marshal()
+	if err != nil {
+		return "", err
+	}
+	return string(res), nil
+}
+
+// Given a string, decodes it into a result (or nil if failed)
+//
+// Sets the metadata to a blank as we don't use it
+// and we don't actually have metadata from M3DB for this result
+func resultDecode(val []byte) (*storage.PromResult, error) {
+	var result storage.PromResult
+	result.PromResult = &prompb.QueryResult{}
+
+	// Check length first so we don't convert to string every time
+	if len(val) == len([]byte(EmptyResult)) && string(val) == EmptyResult {
+		result.Metadata = block.NewResultMetadata()
+		return &result, nil
+	}
+	// Custom Unmarshal function for QueryResults
+	if err := result.PromResult.Unmarshal(val); err != nil {
+		return nil, err
+	}
+	result.Metadata = block.NewResultMetadata()
+	return &result, nil
+}
+
+func (cache *RedisCache) FlushAll() {
+	var response string
+	cache.client.Do(radix.Cmd(&response, "FLUSHALL"))
```

Comment:
Would you add some comments about this method, what data we get from `flushall`?

Reply:
Yeah `flushall` just clears the cache and I just wrote it to for utility in a local unit test, will update with comments